### ISSUE TITLE
relocation diagnostics asserts in response to code that is broken for unrelated reasons #awesome

### ISSUE THREAD

**Issue by ojhunt** [llvm/llvm-project#143325]
I made an error in one of my test cases and got a crash due to a null value. After much confusion I realized the problem is actually that I had redefined the same struct multiple times, and that's what caused the unexpected state.

The test case:

```cpp
struct Foo  {
  Foo(const Foo&);
  ~Foo();
};

struct Foo {
  Foo();
  int;
};

struct Wrapper {
  union {
    Foo p;
  } u;
};
static_assert(__builtin_is_cpp_trivially_relocatable(Wrapper));
```

The assertion is triggered under `DiagnoseNonTriviallyRelocatableReason` in this code:
```cpp
  if (!D->hasSimpleMoveConstructor() && !D->hasSimpleCopyConstructor()) {
    const auto *Decl = cast<CXXConstructorDecl>(
        LookupSpecialMemberFromXValue(SemaRef, D, /*Assign=*/false));
    if (Decl && Decl->isUserProvided())
      SemaRef.Diag(Loc, diag::note_unsatisfied_trait_reason)
          << diag::TraitNotSatisfiedReason::UserProvidedCtr
          << Decl->isMoveConstructor() << Decl->getSourceRange();
  }
```

The issue is the `cast< CXXConstructorDecl >` as `LookupSpecialMemberFromXValue` can return null. The `cast<>` is followed by a null check so it's presumably expected that `LookupSpecialMemberFromXValue` returns null, in which case the fix is likely just to replace `cast` with `cast_or_null`. If it's expected that the cast itself may fail then I assume there's `dyn_cast_or_null` that should be used.

**Comment by ojhunt** [comment]
pre-assigning to @cor3ntin because I blame him for everything, but this is a trivial fix that any one can do if they're wanting to learn how to setup a checkout and build configuration, learn how to add and run tests, etc

**Comment by ojhunt** [comment]
I was looking at my original test case and the duplicate definition failure did not exist, with a bunch of trimming:

```cpp
struct Polymorphic  {
  virtual ~Polymorphic();
};

struct UnionOfPolymorphic {
  union {
    Polymorphic p;
    int i;
  } u;
};

static_assert(__builtin_is_cpp_trivially_relocatable(UnionOfPolymorphic));

```

This triggers the assertion even though there's no error in the code. Note that the static_assert is not necessarily reflecting the actual expected result, but is what is necessary to trigger the diagnostic path.

**Comment by cor3ntin** [comment]
Yeah, I'll look at that today
